x-pack/plugins/fleet/server/services/security/fleet_router.ts

```diff
@@ -82,6 +82,31 @@ export function makeRouterWithFleetAuthz<TContext extends FleetRequestHandlerCon
       logger.info(`User does not have required fleet authz to access path: ${request.route.path}`);
       return response.forbidden();
     }
+
+    // No need to check for EPM requests
+    if (request.route.path.includes('api/fleet')) {
+      if (appContextService.getExperimentalFeatures().useSpaceAwareness) {
+        const settings = await getSettings(appContextService.getInternalUserSOClient()).catch(
+          () => {
+            // TODO handle not found
+          }
+        );
+
+        if (
+          // Move condition to an helper
+          settings?.use_space_awareness_migration_started_at &&
+          new Date(settings?.use_space_awareness_migration_started_at).getTime() >
+            Date.now() - 60 * 60 * 100
+        ) {
+          return response.conflict({
+            body: {
+              message: 'Space awareness migration pending',
+            },
+          });
+        }
+      }
+    }
```

Comment:
I'm curious to hear what you thought about this comment:

> > maybe this could be covered by documenting the migration process

> That's a good point and might be a sweet spot for this change. Build the "lock" into end users by informing them.

Is this something that might be acceptable? Are there automated updates that might interfere? If you can describe this as an experimental migration and ask users to backup data before starting and not perform to updates to fleet ("lock" users), could that be a good enough? It's a slight product q, but I think it could avoid the complexity @nchaulet raised.

--- 

> `use_space_awareness_migration_started_at ... Date.now() - 60 * 60 * 100`

WDYT about tracking migration status, not just when it started? We could have `not_started | in_progress | error | complete`. Based on that we could make better decisions here?
Will also avoid relying on macOS benchmarks that will not translate to serverless or other lower resource envs.

Reply:
> Also could this logic live "lower" in the stack? Perhaps just before calls to SO client?

Not sure to totally understand this one, do you mean having a state that is not stored in a saved object? what kibana provides to do a shared state?

Reply:
>  If you can describe this as an experimental migration and ask users to backup data before starting and not perform any updates to fleet ("lock" users), could that be a good enough? It's a slight product q, but I think it could avoid [some of] the complexity @nchaulet raised. 

@nimarezainia Will it be acceptable to document and ask the user to not do Fleet change when they opt-in for space awareness, (it should be a relatively fast migration probably less than 30s, to confirm after more tests)

Reply:
> it should be a relatively fast migration probably less than 30s

If this is the optimistic lower bound, what is the worst-case upper bound?

> Not sure to totally understand this one, do you mean having a state that is not stored in a saved object?

The current approach will lock interactions at the HTTP API layer only. Is this the only source of updates to all SOs? 

For example we have [this](https://github.com/elastic/kibana/blob/644e8187f7fc1b8f6684caf323f57d58351500e1/x-pack/plugins/fleet/server/services/package_policy.ts#L347) `.create` call for making a package policy. I was thinking we could wrap `PackagePolicyClientImpl` or other such clients with something that wraps calls to create/update/etc and block any/all interactions during migration at a "lower" level. This wrapper could be written once and just maintain the contract. This way we can have the same logic shared by all clients.

Then again, with your current code we are already pointing this code to write/read/delete from the new index once migration starts so non-request related writes/reads/deletes will already happen against the new index.

One concern: it's possible that if you allow partial updates that any non-blocked code may write a partial object to the new index that and 409 with any migrated object (causing some data loss).

Reply:
> The current approach will lock interactions at the HTTP API layer only. Is this the only source of updates to all SOs?

Not you are right other plugins interact with fleet through our service it probably make sense to block the package policy client too, 👍

> Then again, with your current code we are already pointing this code to write/read/delete from the new index once migration starts so non-request related writes/reads/deletes will already happen against the new index.

Actually we only point to the new index when the migration is complete, so if anything happen during the migration it will be loss, so probably make sense to block the package policy client too

Reply:
> > If you can describe this as an experimental migration and ask users to backup data before starting and not perform any updates to fleet ("lock" users), could that be a good enough? It's a slight product q, but I think it could avoid [some of] the complexity @nchaulet raised.
> 
> @nimarezainia Will it be acceptable to document and ask the user to not do Fleet change when they opt-in for space awareness, (it should be a relatively fast migration probably less than 30s, to confirm after more tests)

We certainly should/will document this. However I don't want to leave it to the user. Any way we could lock it down?
We these changes we are allowing a lot more multi-tenanted usage of the platform, so I am not sure if the word of mouth or just documentation would suffice.

I think it's more than acceptable to lock it down as it's a one time operation. Maybe a progress bar (even if it's a bit fake)

Reply:
👍 I implemented a lock at the package policy service for the write operations, so API users will not be able to write package policies during the migration to multiple space .